### DISCUSSION TITLE
Integer constants + how to use atan2

### DISCUSSION THREAD

**Discussion by benlabbeus** [MilesCranmer/PySR#344] [Q&A] [answered]
Hi all!

1. Is there a way to constrain the constant to only use specific integers? In my case, 1 and 2.

2. Is there also a way to use a atan2 function? It seems that Julia doesn't have this.

Much appreciated!

**Comment by MilesCranmer** [comment, marked as answer]
1. Yes, the trick for this is to avoid constants altogether, and give `1` and `2` as **features**. For example:

```python
data_cols = 5
integer_cols = 2
X = np.random.randn(100, data_cols + integer_cols)  # Normal data is first 5 columns
X[:, data_cols+0] = 1
X[:, data_cols+1] = 2
```

We have generated some random data and then set the last two columns equal to the constants we wish for PySR to use.

Then, we need to disable PySR from finding its own constants, by making them prohibitively expensive (and also disabling the constant optimization routine to speed things up):

```python
settings_to_disable_constants = dict(
    complexity_of_constants=100,  # Make constants too expensive to use
    should_optimize_constants=False,  # =speed up
    weight_mutate_constant=0.0  # =speed up
)

model = PySRRegressor(**your_settings, **settings_to_disable_constants)
model.fit(X, y, variable_names=["x1", "x2", "x3", "x4", "x5", "_1", "_2"])
```

The equation will show `_1` and `_2` as features, but they are integers. This is just to prevent SymPy from eagerly simplifying things.

2.

Julia's `atan` can be used as a binary operator, in which case it is the same as typical `atan2`. i.e., `atan(x)` is normal `atan`, and `atan(x, y)` is `atan2`.

Hope this helps!
Miles

**Comment by benlabbeus** [comment]
Amazing! Thanks for the prompt reply Miles!

Just wanted to say this is an amazing tech you've built!

**Reply by MilesCranmer** [reply]
Thanks! 🙂

**Comment by benlabbeus** [comment]
Again, really appreciate your help!

I can't seem to get atan2 working, I have the following:

`
binary_operators=["*", "+", "-", "/", "^", "greater", "atan2"],

extra_sympy_mappings={"atan2": lambda x, y: sympy.atan(x,y)},
`

and am getting the error:


`julia.core.JuliaError: Exception 'UndefVarError: `atan2` not defined' occurred while calling julia code:
[*, +, -, /, ^, greater, atan2]
`

**Comment by benlabbeus** [comment]
Nvm! Got it working:

binary_operators=["*", "+", "-", "/", "^", "greater", "atan2(x, y) = atan(x, y)"],
